internal/impl/devbox.go

```diff
@@ -402,36 +405,150 @@ func (d *Devbox) saveCfg() error {
+		}
+		for _, s := range userSvcs {
+			svcSet[s.Name] = s
+		}
+	}
+
+	return svcSet, nil
+
 }
 
 func (d *Devbox) StartServices(ctx context.Context, serviceNames ...string) error {
+	processComposePaths := map[string]struct{}{}
 	if !IsDevboxShellEnabled() {
 		return d.RunScript("devbox", append([]string{"services", "start"}, serviceNames...))
 	}
-	return services.Start(ctx, d.mergedPackages(), serviceNames, d.projectDir, d.writer)
+
+	if !services.ProcessManagerIsRunning(processComposePidfile) {
+		fmt.Println("Process-compose is not running. Starting it now...")
+		fmt.Println("\nTip: We recommend using `devbox services up` to start process-compose and your services")
+		return d.StartProcessManager(ctx, serviceNames, true, "")
+	}
+
+	svcSet, err := d.Services()
+	if err != nil {
+		return err
+	}
+
+	if len(svcSet) == 0 {
+		return usererr.New("No services found in your project")
+	}
+
+	for _, s := range serviceNames {
+		if _, ok := svcSet[s]; !ok {
+			return usererr.New(fmt.Sprintf("Service %s not found in your project", s))
+		} else {
+			if svcSet[s].ProcessComposePath != "" {
+				processComposePaths[svcSet[s].ProcessComposePath] = struct{}{}
+			}
+		}
+	}
+
+	for _, s := range serviceNames {
+		err := services.StartServices(ctx, d.writer, s, d.projectDir)
+		if err != nil {
+			fmt.Printf("Error starting service %s: %s", s, err)
+		} else {
+			fmt.Printf("Service %s started successfully", s)
+		}
+	}
+	return nil
+}
+
+func (d *Devbox) StopServices(ctx context.Context, serviceNames ...string) error {
+	if !IsDevboxShellEnabled() {
+		return d.RunScript("devbox", append([]string{"services", "stop"}, serviceNames...))
+	}
+
+	if !services.ProcessManagerIsRunning(processComposePidfile) {
+		return usererr.New("Process manager is not running. Run `devbox services up` to start it.")
+	}
+
+	if len(serviceNames) == 0 {
+		return services.StopProcessManager(ctx, d.writer, processComposePidfile)
+	}
+
+	svcSet, err := d.Services()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range serviceNames {
+		if _, ok := svcSet[s]; !ok {
+			return usererr.New(fmt.Sprintf("Service %s not found in your project", s))
+		}
+		services.StopServices(ctx, s, d.projectDir, d.writer)
+	}
+	return nil
+}
+
+func (d *Devbox) RestartServices(ctx context.Context, serviceNames ...string) error {
+	if !IsDevboxShellEnabled() {
+		return d.RunScript("devbox", append([]string{"services", "restart"}, serviceNames...))
+	}
+
+	if !services.ProcessManagerIsRunning(processComposePidfile) {
+		fmt.Println("Process-compose is not running. Starting it now...")
+		fmt.Println("\nTip: We recommend using `devbox services up` to start process-compose and your services")
+		return d.StartProcessManager(ctx, serviceNames, true, "")
+	}
+
+	// TODO: Restart with no services should restart the _currently running_ services. This means we should get the list of running services from the process-compose, then restart them all.
+
+	svcSet, err := d.Services()
```

Comment:
Similar to `start` and `stop`, we could fully restart the process manager if you run `devbox services restart` without passing any service names. This would be a bit more complicated though -- we'd probably want to get the list of services that process-compose currently is running, and make sure that only those start when you re-launch process-compose?

Reply:
This would do some weird stuff (see comment above). Process-compose seems to bake the available services in when you start it, so if we did the, `devbox services restart` would have to restart _all_ the services plus the one the user requested. 

This might be doable once I implement `devbox services ls`